경북대 BE_김진현 Step 2 - 관리자 화면

kimjhyun0627 opened the pull request:
## ✨feat: 선물하기 상품 관리자 화면 구현
### 구현 기능: 선물하기 상품 관리가 가능한 관리자 화면 구현
---
#### 상품 목록 조회 (GET /admin/products):
- 모든 상품을 테이블 형태로 표시하는 관리자 화면을 구현했습니다.
- API 서버(ProductController)에서 RestTemplate을 통해 상품 목록을 가져와 Thymeleaf로 렌더링합니다.
- API 호출 실패 시 사용자에게 오류 메시지를 표시하고 빈 목록을 반환합니다.
### 상품 추가 폼 및 처리 (GET /admin/products/new, POST /admin/products):
- 새로운 상품 정보를 입력받는 폼(admin/product_form.html)을 제공합니다.
- 폼에서 사용자가 직접 상품 ID를 입력할 수 있도록 구현했습니다.
- 폼 제출 시 RestTemplate을 통해 /api/products API로 POST 요청을 보내 상품을 추가합니다.
- 상품 추가 성공 시 목록 페이지로 리다이렉트되고, 유효성 검증 실패(400 Bad Request), ID 충돌(409 Conflict), 또는 API 오류 발생 시 폼 페이지로 돌아가 사용자에게 적절한 오류 메시지를 표시합니다.
### 상품 수정 폼 및 처리 (GET /admin/products/{id}/edit, PUT /admin/products/{id}):
- 특정 상품의 기존 정보가 미리 채워진 수정 폼을 제공합니다.
- GET 요청 시 RestTemplate을 통해 /api/products/{id} API에서 기존 상품 정보를 가져와 폼에 바인딩합니다. 상품을 찾을 수 없거나 API 호출 실패 시 목록 페이지로 리다이렉트하며 오류를 알립니다.
- 폼 제출 시 HiddenHttpMethodFilter와 _method 필드를 활용하여 RestTemplate을 통해 /api/products/{id} API로 PUT 요청을 보내 상품을 수정합니다.
- 상품 수정 성공 시 목록 페이지로 리다이렉트되고, 유효성 검증 실패(400 Bad Request), 상품 없음(404 Not Found), 또는 API 오류 발생 시 폼 페이지로 돌아가 사용자에게 적절한 오류 메시지를 표시합니다.
### 상품 삭제 (DELETE /api/products/{id} - 클라이언트 측에서 직접 호출):
- 상품 목록 화면에서 "삭제" 버튼 클릭 시 JavaScript (fetch API)를 사용하여 /api/products/{id} API로 DELETE 요청을 직접 보냅니다.
- 삭제 성공 시 페이지를 새로 고침하여 업데이트된 목록을 표시하고, 실패 시 사용자에게 알림 메시지를 제공합니다. (별도의 AdminProductController 메서드는 없음)

--- 
### ✅ 테스트
AdminProductControllerTest에 상품 목록 조회, 추가, 수정 기능에 대한 모의(Mock) 테스트를 추가하여 컨트롤러와 RestTemplate의 상호작용 및 뷰 렌더링을 검증했습니다. 삭제 기능은 JavaScript 호출로 구현해 따로 테스트 코드를 작성하지 못했습니다.

kimjhyun0627 committed:
docs: add feature list of step2 to README.md

kimjhyun0627 committed:
feat: 관리자 상품 목록 조회 페이지 구현

- 관리자 상품 목록 조회 페이지 구현
- 관련 테스트 코드 작성

kimjhyun0627 committed:
feat: 관리자 상품 추가 폼 및 POST 요청 처리 구현

- 관리자 상품 추가 폼 및 POST 요청 처리 구현
- 관련 테스트 코드 작성
- README 업데이트

kimjhyun0627 committed:
refactor: product_list.html 더미 데이터 삭제

- product_list.html 더미 데이터 삭제

kimjhyun0627 committed:
feat: 관리자 상품 수정 폼 및 PUT 요청 처리 구현

- 관리자 상품 수정 폼 및 POST 요청 처리 구현
- 기존 상품 추가 폼 재사용
- 관련 테스트 코드 작성
- README 업데이트

kimjhyun0627 committed:
feat: 관리자 상품 삭제 기능 구현

- 관리자 상품 삭제 기능 구현
- 수정 및 삭제 버튼 CSS 적용
- README 업데이트

liquidjoo commented on src/main/java/gift/Application.java:
```diff
@@ -2,10 +2,17 @@
+	public RestTemplate restTemplate() {
+		return new RestTemplate();
+	}
 }
```
파일 마지막에 엔터(개행문자)를 넣어주세요!!
이유는 IEEE가 책정한 POSIX에 정의되어 있기 때문인데요~
좀 더 알고 싶으시면 재미삼아 아래의 링크를 보시는 것도 추천 드립니다.
https://velog.io/@doondoony/posix-eol

Intellij 를 사용하실 경우엔
Preferences -> Editor -> General -> Ensure line feed at file end on save 를 체크해주시면
파일 저장 시 마지막에 개행문자를 자동으로 넣어줍니다!

kimjhyun0627 replied:
앗  사실 제가 step1까지 진행 후 노트북을 바꿔서 기존 셋업이 누락된것같습니다..!
수정해두었습니다 !

liquidjoo commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,208 @@
+                    null,
+                    new ParameterizedTypeReference<List<Product>>() {
+                    }
+            );
```
데이터를 가져오는 행위가 될 수 있을 것 같아요 :)
Contoller 에 있는 것도 좋지만 비슷한 역할을 하는 곳이 있을 것 같은데요
비슷한 역할에 대해서 고민해시고 어떻게 처리하는 것이 좋을까요?

kimjhyun0627 replied:
repository와 service를 분리했습니다 !

liquidjoo commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,208 @@
+            List<Product> products = response.getBody();
+            if (products == null) {
+                products = Collections.emptyList();
+            }
```
위의 코멘트가 처리가 된다면 이러한 책임들은 적절한 역할로 옮겨질 것 같네요

liquidjoo commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,208 @@
+                    HttpMethod.POST,
+                    request,
+                    Product.class
+            );
```
위와 코멘트의 케이스라고 생각합니다

liquidjoo commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    @GetMapping("/products/{id}/edit")
+    public String showEditProductForm(@PathVariable Long id, Model model) {
```
예외처리로 인해 코드가 많이 복잡해보이는데요 Controller 에서 모든 것을 처리하는 것보다 별도로 처리할 수 있도록 만들어보면 어떨까요?
이런 코드는 어떤 단점이 있을 수 있을까요?

kimjhyun0627 replied:
에러 핸들러 처리를 시도했는데, 아직 미숙해 코드가 더 지저분해진 것 같습니다...

liquidjoo commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    @PutMapping("/products/{id}")
+    public String updateProduct(@PathVariable Long id, @ModelAttribute Product product,
```
위의 코멘트와 동일합니다 :)

liquidjoo commented on src/main/resources/templates/admin/product_list.html:
```diff
@@ -0,0 +1,149 @@
+</table>
+
+<script>
+  function deleteProduct(productId) {
+    if (confirm('정말로 이 상품을 삭제하시겠습니까?')) {
+      fetch(`/api/products/${productId}`, {
+        method: 'DELETE'
+      })
+      .then(response => {
+        if (response.ok) {
+          alert('상품이 성공적으로 삭제되었습니다.');
+          window.location.reload();
+        } else if (response.status === 404) {
+          alert('삭제할 상품을 찾을 수 없습니다.');
+        } else {
+          alert('상품 삭제에 실패했습니다. 상태 코드: ' + response.status);
+        }
+      })
+      .catch(error => {
+        console.error('삭제 중 오류 발생:', error);
+        alert('상품 삭제 중 오류가 발생했습니다.');
+      });
+    }
+  }
```
삭제는 어떻게 테스트코드를 만들어보면 좋을까요?

liquidjoo replied:
<img width="361" alt="스크린샷 2025-06-26 오후 6 19 30" src="https://github.com/user-attachments/assets/32f2cb45-30b3-4212-b10d-bbc678bba172" />
테스트 코드가 일부 깨져서 보입니다!

liquidjoo reviewed (changes requested):
안녕하세요 진현님 :)
2단계 미션 구현 잘해주셨어요 👍 
몇 가지 생각해보면 좋을 것 같은 코멘트 남겼습니다
피드백 반영해서 다시 요청해 주세요 🙇

kimjhyun0627 committed:
refactor: 역할 분리 및 테스트 개선

- repository, service, controller 역할 분리
- 예외 처리
- POSIX 개행문자 수정

woowahan-justin commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,189 @@
+            return "redirect:/admin/products";
+        } catch (HttpClientErrorException e) {
+            String errorMessage = mapApiErrorToUserMessage((HttpStatus) e.getStatusCode());
+            model.addAttribute("errorMessage", errorMessage);
+            model.addAttribute("product", product);
+            model.addAttribute("formAction", "/admin/products");
+            model.addAttribute("pageTitle", "새 상품 추가");
+            return "admin/product_form";
+        } catch (ResourceAccessException e) {
+            model.addAttribute("errorMessage", "API 서버에 연결할 수 없습니다. 서버가 실행 중인지 확인하세요.");
+            model.addAttribute("product", product);
+            model.addAttribute("formAction", "/admin/products");
+            model.addAttribute("pageTitle", "새 상품 추가");
+            return "admin/product_form";
+        } catch (Exception e) {
+            model.addAttribute("errorMessage", "상품 추가 중 예상치 못한 오류가 발생했습니다: " + e.getMessage());
+            model.addAttribute("product", product);
+            model.addAttribute("formAction", "/admin/products");
+            model.addAttribute("pageTitle", "새 상품 추가");
```
객체를 활용해 보아도 좋을 것 같아요!

woowahan-justin commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,189 @@
+        }
+    }
+
+    @GetMapping("/products/{id}/edit")
```
<img width="1486" alt="스크린샷 2025-06-28 오전 8 37 55" src="https://github.com/user-attachments/assets/b1bf97fa-6a96-48e2-92fc-dbb4e42f2949" />

<img width="1257" alt="스크린샷 2025-06-28 오전 8 38 02" src="https://github.com/user-attachments/assets/e42d96a6-f99e-42ac-80f5-5435aa11667a" />

수정이 작동이 안되는 것 같은데요! 확인 가능할까요?

woowahan-justin commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -21,89 +19,41 @@
         return new ResponseEntity<>(updatedProduct, HttpStatus.OK);
     }
 
     @DeleteMapping("/products/{id}")
```
<img width="1459" alt="스크린샷 2025-06-28 오전 8 38 10" src="https://github.com/user-attachments/assets/15942909-1aec-4adb-9056-8ce5518a0853" />
해당 부분도 확인이 가능할까요?

woowahan-justin replied:
저의 컴퓨터에서 문제가 있을수도 있으나 혹시 확인이 된다면 다시 알려주시면 좋을 것 같아요!

woowahan-justin reviewed (changes requested):
안녕하세요 진현님
피드백 반영 잘해주셨어요!
몇 가지 고민해볼 코멘트 남겼습니다
확인하고 다시 요청해 주세요 🙇

kimjhyun0627 committed:
refactor: 예외 처리 객체 분리

- 예외 처리를 위한 객체 생성 및 분리

talk about: 여러 방식으로 테스트해봤는데 제 개발환경에서는 정상 작동하고 있습니다. 수정 및 삭제가 제대로 이루어지지 않는 상황을 자세히 알려주시면 재확인해보겠습니다..!

kimjhyun0627 commented:
어제 밤늦게 리뷰요청 올렫려 정말 죄송합니다! 워크샵 일정 상 최대한 빨리 작성해보려고 했지만...ground rule을 간과했던 것 같습니다.
추가로, 오류 관련 코멘트를  작성했었는데, pending 상태로 종료하여 삭제된 것 같습니다...ㅜㅜ
이 부분 추가해서 코멘트 남기겠습니다 !

liquidjoo commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    @GetMapping("/products/{id}/edit")
```
스프링에서 명시적인 리플렉션 보안 제한으로 인해 @RequestParam, @PathVariable, @RequestBody 에 대한 이름을 명시적으로 적어주어야 합니다 :)
제가 임의로 명시적으로 작성 후 실행을 해보았는데요 
`/admin/products?error=상품 정보를 가져오는 중 API 오류가 발생했습니다: 400] has been removed from the response because it is invalid`

이렇게 에러가 나오네요!! 해당 부분 수정 후 다시 요청 부탁드립니다

liquidjoo commented on src/main/java/gift/exception/ApiExceptionHandler.java:
```diff
@@ -0,0 +1,38 @@
+        String errorMessage = e.getMessage();
+        HttpStatus status = HttpStatus.BAD_REQUEST; // 400 Bad Request
+
+        if (errorMessage.contains("상품을 찾을 수 없습니다")) {
+            status = HttpStatus.NOT_FOUND; // 404 Not Found
+        } else if (errorMessage.contains("이미 존재하는 상품 ID")) {
+            status = HttpStatus.CONFLICT; // 409 Conflict
+        } else if (errorMessage.contains("유효하지 않습니다")) {
+            status = HttpStatus.BAD_REQUEST; // 400 Bad Request
+        } else if (errorMessage.contains("ID가 일치하지 않습니다")) {
+            status = HttpStatus.BAD_REQUEST; // 400 Bad Request
+        }
```
if else 를 최소화 할 수 있는 방안은 어떤 것들이 있을까요?

liquidjoo commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class ProductServiceImpl implements ProductService {
```
다중 if 를 최소화 하면 어떨까요?

liquidjoo commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+        productRepository.deleteById(id);
+    }
+
+    private boolean validateProduct(Product product) {
+        if (product.getName() == null || product.getName().trim().isEmpty()) {
+            System.err.println("유효성 검증 실패: 상품 이름은 필수 입력 항목입니다.");
+            return false;
+        }
+        if (product.getPrice() <= 0) {
+            System.err.println("유효성 검증 실패: 상품 가격은 0보다 커야 합니다.");
+            return false;
+        }
+        String imageUrl = product.getImageUrl();
+        if (imageUrl == null || imageUrl.trim().isEmpty()) {
+            System.err.println("유효성 검증 실패: 상품 이미지 URL은 필수 입력 항목입니다.");
+            return false;
+        }
+        String urlRegex = "^(https?|ftp)://[^\\s/$.?#].[^\\s]*$";
+        if (!imageUrl.matches(urlRegex)) {
+            System.err.println("유효성 검증 실패: 상품 이미지 URL 형식이 올바르지 않습니다.");
+            return false;
+        }
+        return true;
+    }
```
이런 유효성 검사는 Service 의 영역에서 처리하는 것이 좋을까요?
그렇다면 Product 객체는 단순히 데이터만 가져오고 만들어주는 역할만 가질까요?
고민해보면 좋을 것 같아요

liquidjoo reviewed (changes requested):
안녕하세요 진현님 :)
피드백 반영 잘해주셨어요
좀 더 생각해볼만한 코멘트들 남겼습니다
확인하고 다시 요청해 주세요 🙇

kimjhyun0627 committed:
refactor: 전체 파일 구조 리팩토링 및 수정

- 전반적인 프로젝트 구조 변경
- 상품 ID를 서버사이드에서 라벨링
- 예외 처리 로직 재구성
- 테스트 코드 로직 변경

kimjhyun0627 commented:
말씀해주신 부분들 바탕으로 많은 생각을 해 보았습니다...
(아직까지 왜 제 개발환경에서는 오류가 발생하지 않았는지 모르겠습니다 ㅜㅜ 동일한 현상이 있었으면 고쳤을 것 같기도 하고, 이번 커밋도 동일한 증상이 발생할 가능성이 있으니...)
결론적으로 제 실력에 비해 전반적인 프로젝트 구조를 너무 안일하게 조직했다는 생각이 들었고
기존 코드를 최대한 살리되 조금 더 보편이고 범용적인 구조로 다시 프로젝트 및  구현 로직을 구성했습니다.
- 전반적인 프로젝트 구조 변경
- 상품 ID를 서버사이드에서 라벨링하게 변경 → 409와 같은 불필요한 오류처리 최소화
- 예외처리 구조 재구성
- 3단계 과제 역시 딜레이됨에 따라, 최대한 빠르게 3단계를 끝낼 수 있게 함수명을 JPA에 가깝게 통일
- 테스트 코드 로직 변경
 → 테스트 코드의 경우...금일 오전 특강에서 테스트 코드 관련 내용을 접한지라, 기존 코드가 많이 잘못되어 있었음을 알게 되었습니다.
아직은 많이 부족하다는 것을 알지만, 코맨트 달아주시면 감사하겠습니다.

liquidjoo commented on src/main/java/gift/controller/admin/AdminProductController.java:
```diff
@@ -0,0 +1,71 @@
+
+
+    @GetMapping("/{id}/edit")
+    public String editForm(@PathVariable("id") Long id, Model model) {
```
`@PathVaribale` 를 사용하는 경우 스프링 보안 강화에 따라서 명시적으로 표현해주는 것이 좋습니다
피드백 반영 좋아요!

liquidjoo commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -0,0 +1,46 @@
+    public ProductRequest() {
+    }
+
+    // getters / setters
```
record class 도 한번 고려해보면 좋을 것 같아요
record class 는 어떤 역할을 가지고 있는지 찾아보면 좋습니다

liquidjoo commented on src/main/java/gift/entity/Product.java:
```diff
@@ -0,0 +1,54 @@
+        this.imageUrl = imageUrl;
+    }
+
+    // getters / setters
```
주요 비즈니스를 담당할 수 있는 Product 에는 setter 보다는 명시적인 이름을 표현해보면 어떨까요?
이러한 객체에 setter 를 사용한다면 어떤 일들이 일어날 수 있는지 고민해 보아요!

liquidjoo reviewed (approved):
안녕하세요 진현님
피드백 반영 잘해주셨어요 👍 
몇 가지 코멘트 남겼는데요~ 확인하고 반영해보면 좋을 것 같아요
다음 단계 진행해 주세요 🙇
